arraycontext/pytest.py

```diff
@@ -28,48 +29,124 @@
 THE SOFTWARE.
 """
 
+from typing import Optional, Sequence
 
-# {{{ pytest integration
+import pyopencl as cl
+from pyopencl.tools import _ContextFactory
 
-def pytest_generate_tests_for_pyopencl_array_context(metafunc):
-    """Parametrize tests for pytest to use a
-    :class:`~arraycontext.PyOpenCLArrayContext`.
 
-    Performs device enumeration analogously to
-    :func:`pyopencl.tools.pytest_generate_tests_for_pyopencl`.
+# {{{ array context factories
 
-    Using the line:
+class _PyOpenCLArrayContextFactory(_ContextFactory):
+    force_device_scalars = True
+
+    def __call__(self):
+        ctx = super().__call__()
+        from arraycontext.impl.pyopencl import PyOpenCLArrayContext
+        return PyOpenCLArrayContext(
+                cl.CommandQueue(ctx),
+                force_device_scalars=self.force_device_scalars)
+
+    def __str__(self):
+        return ("<array context factory for <pyopencl.Device '%s' on '%s'>" %
+                (self.device.name.strip(),
+                 self.device.platform.name.strip()))
+
+
+class _DeprecatedPyOpenCLArrayContextFactory(_PyOpenCLArrayContextFactory):
+    force_device_scalars = False
+
+
+_ARRAY_CONTEXT_FACTORY_DICT = {
+        "pyopencl": _PyOpenCLArrayContextFactory,
+        }
+_ALL_ARRAY_CONTEXT_FACTORY_DICT = {
+        "pyopencl-deprecated": _DeprecatedPyOpenCLArrayContextFactory,
+        }
+
+_ALL_ARRAY_CONTEXT_FACTORY_DICT.update(_ARRAY_CONTEXT_FACTORY_DICT)
+
+# }}}
+
+
+# {{{ pytest integration
+
+def pytest_generate_tests_for_array_context(
+        metafunc,
+        impls: Optional[Sequence[str]] = None):
+    """Parametrize tests for pytest to use an :class:`~arraycontext.ArrayContext`.
+
+    Using the line
 
     .. code-block:: python
 
-       from arraycontext import pytest_generate_tests_for_pyopencl
-            as pytest_generate_tests
+       from arraycontext import (
+            pytest_generate_tests_for_array_context
+            as pytest_generate_tests)
 
     in your pytest test scripts allows you to use the argument ``actx_factory``,
-    in your test functions, and they will automatically be
-    run once for each OpenCL device/platform in the system, as appropriate,
-    with an argument-less function that returns an
-    :class:`~arraycontext.ArrayContext` when called.
+    which is a callable that returns a :class:`~arraycontext.ArrayContext`.
+    All test functions will automaticall be run once for each implemented array
+    context. To select specific array context implementations explicitly
+    define, for example,
 
-    It also allows you to specify the ``PYOPENCL_TEST`` environment variable
-    for device selection.
+    .. code-block:: python
+
+        def pytest_generate_tests(metafunc):
+            pytest_generate_tests_for_array_context(metafunc, impls=["pyopencl"])
```

Comment:
I think this can be a function that returns a function/closure, so that this type of thing would work:
```py
pytest_generate_tests = pytest_generate_tests_for_array_contexts(["pyopencl"])
```

Reply:
Made this into a closure in 80d5e4b.